### PR TITLE
feat(frontdoor-agw): Add implementation for Azure Frontdoor and App Gateway inferred spans 

### PR DESCRIPTION
**Pull request by TophrC-dd**
### What does this PR do?
This PR implements Azure FrontDoor and Azure Application Gateway. These are upstream proxy services Azure offers.

**[Azure Front Door](https://azure.microsoft.com/en-us/products/frontdoor)** 
Azure Front Door is a edge computing CDN proxy that provides load balancing and regional routing to azure web apps and containers. Customers use this ensure their application has a single entry point across multiple regions. 

image

**[Azure Application Gateway](https://learn.microsoft.com/en-us/azure/application-gateway/overview)** 
Azure Application Gateway is upstream proxy service designed for complex routing for internal applications. This proxy gets placed within private networks. 

image

Both of these use the existing code paths for other upstream proxies. However both of these services lack the ability to insert timestamps in the headers. This may change in the future but due to this I modified the process to add a timespan at header detection which means that the accuracy of the timeline decreases but we still get a usable span and the proxy is represented. If this does change and the ability to insert timestamps does get implemented on Azure's side the code is thoughtful enough to not overwrite an existing timestamp header and making it less of an urgent change to remove the code I am introducing. 

### Motivation
There maybe future customer interest in having observability into these products. Getting these in quickly as possible will provide the benefit of having that ready when that demand rises. 

### Additional Notes
In my testing use the existing inferred proxy tests that include a timestamp. For completeness I also built a test where the timestamp is missing from the headers to ensure one is generated. Azure does  provide a timestamp for front door with the header `x-azure-ref` but the precision is down to the the second and not millisecond. I think generating the timestamp is better as having a full second of a difference either way could look worse in span graphs.

### CODE REVIEW

File: packages/dd-trace/src/plugins/util/inferred_proxy.js [hunk 1]
```diff
@@ -109,19 +122,21 @@ function setInferredProxySpanTags (span, proxyContext) {
...
+
+  if (detectedProxy.providesTimestamp && !(PROXY_HEADER_START_TIME_MS in headers)) {
     return null
   }
 
   return {
     requestTime: headers[PROXY_HEADER_START_TIME_MS]
       ? Number.parseInt(headers[PROXY_HEADER_START_TIME_MS], 10)
-      : null,
+      : Date.now().toString(),
```
Comment (BridgeAR) [inline packages/dd-trace/src/plugins/util/inferred_proxy.js]: Please keep this as number to align with the former behavior

```suggestion
      : Date.now(),
```

File: packages/dd-trace/test/plugins/util/inferred_proxy.spec.js [hunk 2]
```diff
@@ -333,6 +391,83 @@ Object.entries(proxyConfigs).forEach(([proxyType, config]) => {
...
+              'http.method': 'GET',
+              'http.status_code': '200',
+              component: testCase.expectedComponent,
+              '_dd.integration': testCase.expectedComponent,
+            },
+            metrics: {
+              '_dd.inferred_span': 1,
+            },
+          })
+          assert.notStrictEqual(spans[0].start.toString(), null)
```
Comment (BridgeAR) [inline packages/dd-trace/test/plugins/util/inferred_proxy.spec.js]: Could we just mock the time above and verify it here? :)

That way it is a stronger test and I believe the actual testing part here right now is that start is not nullish as before (if it would be, it would actually already throw at the `.toString()`)

File: packages/dd-trace/src/plugins/util/inferred_proxy.js [hunk 3]
```diff
@@ -109,19 +122,21 @@ function setInferredProxySpanTags (span, proxyContext) {
...
+  if (!(PROXY_HEADER_SYSTEM in headers && headers[PROXY_HEADER_SYSTEM] in supportedProxies)) {
+    log.debug('Received headers to create inferred proxy span but headers include an unsupported proxy type', headers)
     return null
   }
 
-  if (!(PROXY_HEADER_SYSTEM in headers && headers[PROXY_HEADER_SYSTEM] in supportedProxies)) {
-    log.debug('Received headers to create inferred proxy span but headers include an unsupported proxy type', headers)
+  const detectedProxy = supportedProxies[headers[PROXY_HEADER_SYSTEM]]
+
+  if (detectedProxy.providesTimestamp && !(PROXY_HEADER_START_TIME_MS in headers)) {
```
Comment (BridgeAR) [inline packages/dd-trace/src/plugins/util/inferred_proxy.js]: This is an edge case where I can not say much about: if PROXY_HEADER_START_TIME_MS is present but empty, it used to produce `null`. Now, if `detectProxy.providesTimestamp` is false, it creates Date.now(). Is that intentional? :)

File: packages/dd-trace/test/plugins/util/inferred_proxy.spec.js (hunk 2 above)
Comment (TophrC-dd) [reply packages/dd-trace/test/plugins/util/inferred_proxy.spec.js]: I will strengthen the mock testing to address the issue regarding the edge case of the empty timestamp. It may resolve to 0 and that wouldn't be good.

File: packages/dd-trace/src/plugins/util/inferred_proxy.js (hunk 3 above)
Comment (TophrC-dd) [reply packages/dd-trace/src/plugins/util/inferred_proxy.js]: It's intentional for azure frontdoor as there is no native way for it to pass its own request timestamps. I did have a previous commit that looked like this:
`
    requestTime: headers[PROXY_HEADER_START_TIME_MS]
      ? Number.parseInt(headers[PROXY_HEADER_START_TIME_MS], 10)
      : detectedProxy.providesTimestamp
        ? null
        : Date.now().toString(),
`
which I can fix to: 
`
    requestTime: headers[PROXY_HEADER_START_TIME_MS]
      ? Number.parseInt(headers[PROXY_HEADER_START_TIME_MS], 10)
      : detectedProxy.providesTimestamp
        ? null
        : Date.now(),
`

Codecov wasn't happy with my test and this is real-edge case. if it's preferred i can revert to the above code and work on my test units. my Screenshots provided from dd ui are using the .toString values.

**Review by jcstorms1 — DISMISSED** [review]
Serverless approach looks good 👍

**Review by BridgeAR — COMMENTED** [review]
LGTM

The AI comment is legit, while I am on the fence how to handle it. We could have a lower volume one, but I believe it is not critical

**Review by BridgeAR — DISMISSED** [review]

File: packages/dd-trace/src/plugins/util/inferred_proxy.js [hunk 4]
```diff
@@ -109,19 +122,21 @@ function setInferredProxySpanTags (span, proxyContext) {
...
     return null
```
[the remark this answers is not in the document]
Comment (TophrC-dd) [reply packages/dd-trace/src/plugins/util/inferred_proxy.js:138]: I changed the logic in this so that if no proxy header is present it returns null -- if the header is present and its unsupported proxy type it will log the error with just the proxy type name.

File: packages/dd-trace/src/plugins/util/inferred_proxy.js (hunk 3 above)
Comment (TophrC-dd) [reply packages/dd-trace/src/plugins/util/inferred_proxy.js]: I fixed this logic

File: packages/dd-trace/test/plugins/util/inferred_proxy.spec.js [hunk 5]
```diff
@@ -333,6 +391,82 @@ Object.entries(proxyConfigs).forEach(([proxyType, config]) => {
         })
       })
 
+      it('should create a timestamp if it is an Azure Application Gateway proxy', async () => {
```
[the remark this answers is not in the document]
Comment (TophrC-dd) [reply packages/dd-trace/test/plugins/util/inferred_proxy.spec.js:394]: That makes sense

File: packages/dd-trace/src/plugins/util/inferred_proxy.js [hunk 6]
```diff
@@ -109,21 +122,32 @@ function setInferredProxySpanTags (span, proxyContext) {
...
+
+  if (detectedProxy.providesTimestamp && !headers[PROXY_HEADER_START_TIME_MS]) {
     return null
   }
 
   return {
     requestTime: headers[PROXY_HEADER_START_TIME_MS]
       ? Number.parseInt(headers[PROXY_HEADER_START_TIME_MS], 10)
-      : null,
+      : Date.now(),
```
Comment (tlhunter) [inline packages/dd-trace/src/plugins/util/inferred_proxy.js:144]: Something like this would probably alleviate Copilot's concern:

```suggestion
    requestTime: Number.parseInt(headers[PROXY_HEADER_START_TIME_MS], 10) || Date.now(),
```

It would attempt the parse, and if it gets a number us it, and if not it would fallback to the current timestamp.

Comment (TophrC-dd) [reply packages/dd-trace/src/plugins/util/inferred_proxy.js:144]: That makes sense I will apply your suggestion. 

Thank you

**Review by tlhunter — APPROVED** [review]
